client/rpc/trace/src/lib.rs

```diff
@@ -0,0 +1,375 @@
+	RpcError {
+		code: ErrorCode::InternalError,
+		message: message.to_string(),
+		data: None,
+	}
+}
+
+pub type Responder = oneshot::Sender<Result<Vec<TransactionTrace>>>;
+pub type TraceFilterCacheRequester = TracingUnboundedSender<(FilterRequest, Responder)>;
+
+pub struct TraceFilterCache<B, C, BE, A>(PhantomData<(B, C, BE, A)>);
+
+impl<B, C, BE, A> TraceFilterCache<B, C, BE, A>
+where
+	BE: Backend<B> + 'static,
+	BE::State: StateBackend<BlakeTwo256>,
+	BE::Blockchain: BlockchainBackend<B>,
+	C: ProvideRuntimeApi<B> + AuxStore,
+	C: HeaderMetadata<B, Error = BlockChainError> + HeaderBackend<B>,
+	C: Send + Sync + 'static,
+	C: StorageProvider<B, BE>,
+	C: HeaderMetadata<B, Error = BlockChainError> + 'static,
+	B: BlockT<Hash = H256> + Send + Sync + 'static,
+	C::Api: BlockBuilder<B, Error = BlockChainError>,
+	C::Api: DebugRuntimeApi<B>,
+	C::Api: EthereumRuntimeRPCApi<B>,
+	A: EthApi,
+{
+	// TODO :
+	// 1. Handle requests and add traces to the cache :
+	//    Cache is a BTreeMap : Block height => Vec of TransactionTrace + expiration time
+	//    No filtering is done in the cache, it stores all traces.
+	//    Existing block in cache get the expiration time bumped.
+	//
+	//    Filtering is done on top :
+	//    1. Apply the filter and return a list of indices to keep
+	//    2. Use the indices to build the filtered vec of traces (with correct pointers)
+	//
+	// 2. Remove expired cache :
+	//    Iterate over each block in the BTreeMap, and remove the entry if expired.
+	//    Question : Is the expiration time and delay between checks configurable ?
+	//    Other idea : Spawn a timer future when updating the cache, providing the height
+	//        of the block. When woken up, check only this block expiration time.
+	//        Will create a future for each request, but is more reactive to cleanup.
+	//        Which is better ?
+
+	// How to get Ethereum block hash and transactions :
+	// EthApi::block_by_number -> Rich<Block>
+	// Block.transactions (Full).hash
+	// Debug RPC impl shows how to get the substrate equivalents for mapping.
+
+	pub fn task(
+		client: Arc<C>,
+		backend: Arc<BE>,
+		eth_api: A,
+	) -> (impl Future<Output = ()>, TraceFilterCacheRequester) {
+		const EXPIRATION_DELAY: Duration = Duration::from_secs(600);
+
+		let (tx, mut rx): (TraceFilterCacheRequester, _) =
+			sp_utils::mpsc::tracing_unbounded("trace-filter-cache-requester");
+
+		let fut = async move {
+			let mut expiration_futures = FuturesUnordered::new();
+			let mut cached_blocks = BTreeMap::<u32, CacheBlock>::new();
+
+			tracing::info!("Begining Trace Filter Cache Task ...");
+
+			'service: loop {
+				select! {
+					req = rx.next() => {
+						if let Some((req, response_tx)) = req {
+							let span = tracing::debug_span!("received request", request = ?req);
+							let _guard = span.enter();
+
+							tracing::trace!("Begining handling request");
+
+							let from_block = match req.from_block {
+								None => 1,
```

Comment:
When the user does not provide a `from_block` in the request's filter, I believe that the trace should be performed only on the latest block, but we should check how this behaves on OpenEthereum.